### PR TITLE
[inductor] Fix lowerings that create unexpected aliases

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #106319
* __->__ #105173

This may give the wrong result in some cases, e.g.
```python
@torch.compile()
def fn(x):
    tmp = x.ceil()
    x.add_(10)
    return tmp

a = torch.zeros((), dtype=torch.int64)
fn(a)  # tensor(10)
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

### CODE REVIEW

File: torch/_inductor/lowering.py [hunk 1]
```diff
@@ -186,18 +186,19 @@ def transform_args(args, broadcast, type_promotion_kind, convert_input_to_bool):
...
-                new_args.append(to_dtype(args[i], dtype))
-            elif isinstance(args[i], ir.Constant):
-                new_args.append(
-                    ir.Constant(args[i].value, dtype, args[indices[0]].get_device())
-                )
+        def promote(arg):
+            if isinstance(arg, TensorBox):
+                if arg.get_dtype() != dtype:
+                    return to_dtype(arg, dtype)
+                return arg
```
Comment (peterbell10) [inline torch/_inductor/lowering.py:195]: This is the only functional change here. It fixes an infinite recursion issue created by `to_dtype` calling `clone`, which is wrapped with type promotion that itself calls `to_dtype`. This only happens for no-op dtype conversions, where it was wasteful to clone anyway so better to not call `to_dtype` when the dtypes already match here.

**Review by lezcano — APPROVED** [review]

**Comment by chunyuan-w** [comment]
Hi @peterbell10, I tried with your PR but it seems the unit test provided in your PR description still gives incorrect result:
```
import torch

# @torch.compile()
def fn(x):
    tmp = x.ceil()
    x.add_(10)
    return tmp

a = torch.zeros((), dtype=torch.int64)
print(fn(a))  # tensor(10)
```

`print(fn(a))` still gives `tensor(10)` using PyTorch commit https://github.com/pytorch/pytorch/commit/4dc063821dbd9374c7a4dffdfb1508add1d2c367 (which includes your fix PR) while eager mode returns `tensor(0)`. Could you please help double-confirm if this is the case from your side?

In addition, the change in this PR has brought performance regression to several models:
https://github.com/pytorch/pytorch/issues/106962
https://github.com/pytorch/pytorch/issues/106963
https://github.com/pytorch/pytorch/issues/106964

As we're approaching the PyTorch 2.1 branch cut, may I know if you could help take a look at this issue?

**Comment by lezcano** [comment]
I forgot to ask you to add the script in the OP as a test. Could you do so while investigating those issues in afollow up @peterbell10 ?

**Pull request closed without merging**